packages/eventindexer/indexer/filter_then_subscribe.go

```diff
@@ -0,0 +1,112 @@
+package indexer
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/taikoxyz/taiko-mono/packages/eventindexer"
+)
+
+func (svc *Service) FilterThenSubscribe(
+	ctx context.Context,
+	mode eventindexer.Mode,
+	watchMode eventindexer.WatchMode,
+) error {
+	chainID, err := svc.ethClient.ChainID(ctx)
+	if err != nil {
+		return errors.Wrap(err, "svc.ethClient.ChainID()")
+	}
+
+	if watchMode == eventindexer.SubscribeWatchMode {
+		return svc.subscribe(ctx, chainID)
+	}
+
+	if err := svc.setInitialProcessingBlockByMode(ctx, mode, chainID); err != nil {
+		return errors.Wrap(err, "svc.setInitialProcessingBlockByMode")
+	}
+
+	header, err := svc.ethClient.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return errors.Wrap(err, "svc.ethClient.HeaderByNumber")
+	}
+
+	if svc.processingBlockHeight == header.Number.Uint64() {
+		log.Infof("chain ID %v caught up, subscribing to new incoming events", chainID.Uint64())
+		return svc.subscribe(ctx, chainID)
+	}
+
+	log.Infof("chain ID %v getting events between %v and %v in batches of %v",
+		chainID.Uint64(),
+		svc.processingBlockHeight,
+		header.Number.Int64(),
+		svc.blockBatchSize,
+	)
+
+	for i := svc.processingBlockHeight; i < header.Number.Uint64(); i += svc.blockBatchSize {
+		end := svc.processingBlockHeight + svc.blockBatchSize
+		// if the end of the batch is greater than the latest block number, set end
+		// to the latest block number
+		if end > header.Number.Uint64() {
+			end = header.Number.Uint64()
+		}
+
+		filterOpts := &bind.FilterOpts{
+			Start:   svc.processingBlockHeight,
+			End:     &end,
+			Context: ctx,
+		}
+
+		blockProvenEvents, err := svc.taikol1.FilterBlockProven(filterOpts, nil)
```

Comment:
do we need to check `blockProvenEvents.Event.Raw.Removed`?
<img width="856" alt="image" src="https://user-images.githubusercontent.com/104078303/227511943-a5969662-d264-469b-98e8-25fe902fbf55.png">